hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMBucketLayoutUpgrade.java

```diff
@@ -0,0 +1,263 @@
+  private OzoneManagerProtocol omClient;
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(TestOMBucketLayoutUpgrade.class);
+
+  /**
+   * Defines a "from" layout version to finalize from.
+   *
+   * @return
+   */
+  @Parameterized.Parameters
+  public static Collection<Object[]> data() {
+    return Arrays.asList(new Object[][]{
+        {INITIAL_VERSION},
+        {ERASURE_CODED_STORAGE_SUPPORT},
+    });
+  }
+
+
+  public TestOMBucketLayoutUpgrade(OMLayoutFeature fromVersion) {
+    this.fromLayoutVersion = fromVersion.layoutVersion();
+  }
+
+  /**
+   * Create a MiniDFSCluster for testing.
+   */
+  @Before
+  public void setup() throws Exception {
+    org.junit.Assume.assumeTrue("Check if there is need to finalize.",
+        maxLayoutVersion() > fromLayoutVersion);
+
+    OzoneConfiguration conf = new OzoneConfiguration();
+    String omServiceId = UUID.randomUUID().toString();
+    cluster = (MiniOzoneHAClusterImpl) MiniOzoneCluster.newOMHABuilder(conf)
+        .setClusterId(UUID.randomUUID().toString())
+        .setScmId(UUID.randomUUID().toString())
+        .setOMServiceId(omServiceId)
+        .setNumOfOzoneManagers(3)
+        .setNumDatanodes(1)
+        .setOmLayoutVersion(fromLayoutVersion)
+        .build();
+
+    cluster.waitForClusterToBeReady();
+    ozoneManager = cluster.getOzoneManager();
+    ObjectStore objectStore = OzoneClientFactory.getRpcClient(omServiceId, conf)
+        .getObjectStore();
+    clientProtocol = objectStore.getClientProxy();
+    omClient = clientProtocol.getOzoneManagerClient();
+
+    // create sample volume.
+    omClient.createVolume(
+        new OmVolumeArgs.Builder()
+            .setVolume(VOLUME_NAME)
+            .setOwnerName("user1")
+            .setAdminName("user1")
+            .build());
+  }
+
+  /**
+   * Shutdown MiniDFSCluster.
+   */
+  @After
+  public void shutdown() {
+    if (cluster != null) {
+      cluster.shutdown();
+    }
+  }
+
+  /**
+   * Tests that OM blocks all requests to create any buckets with a new bucket
+   * layout.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testCreateBucketWithNewLayoutsPreFinalize() throws Exception {
+    // Assert OM layout version is 'fromLayoutVersion' on deploy.
+    assertEquals(fromLayoutVersion,
+        ozoneManager.getVersionManager().getMetadataLayoutVersion());
+    assertNull(ozoneManager.getMetadataManager().getMetaTable()
+        .get(LAYOUT_VERSION_KEY));
+
+    for (BucketLayout layout : BUCKET_LAYOUTS) {
+      if (layout.isLegacy()) {
+        // We only want to test out the newly introduced layouts in this test.
+        continue;
+      }
+      try {
+        LOG.info("Creating bucket with layout {} during Pre-Finalize", layout);
+        createBucketWithLayout(layout);
+        fail("Expected to fail creating bucket with layout " + layout);
+      } catch (OMException e) {
+        // Expected exception.
+        assertTrue(e.getMessage().contains(
```

Comment:
You can check the result code of `OMException#getResult` for `NOT_SUPPORTED_OPERATION_PRIOR_FINALIZATION` instead of parsing the error message, which might be more robust.

Reply:
Thank you for the suggestion, I have made this change.